### PR TITLE
[JENKINS-30086] Steps help and description clean up

### PR DESCRIPTION
**Pull request by amuniz**
Both related:
[JENKINS-30086](https://issues.jenkins-ci.org/browse/JENKINS-30086)

Work in progress:
- [X] Review display names.
- [X] General inline help.
- [X] Fields inline help (if not clear from the field name).
- [X] Include the step function name in the dropdown list (format: `stepname: Step Display name`).

@reviewbybees (esp. @estheryouhana)

### CODE REVIEW

**Comment by amuniz** [comment]
@jglick It seems that I can not add labels (?) Could you add `work-in-progress` please?

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<p>
```
Comment (recena) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: `div` lost

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help-path.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<p>
+    Current directory (<code>pwd()</code>) relative path to the Groovy file to load.
+</p>
```
Comment (recena) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help-path.html]: Ditto

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html (hunk 1 above)
Comment (amuniz) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: I followed the rule in all other inlined help files for fields (the general help files use `<div>` instead).

Comment (recena) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: All help files should include `<div>` as a wrapper.

Comment (recena) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: There are a lot of anti-examples. [Here](https://github.com/jenkinsci/subversion-plugin/blob/master/src/main/resources/hudson/scm/SubversionSCM/AdditionalCredentials/help-realm.html) one of them.

Comment (amuniz) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: I have no strong opinion, but what is the difference? I mean, there is a `div` already added by core, so adding `p` inside `div` is fine. The final HTML output is this:

```html
<div class="help" style="display: block;">
  <p>
    Every <code>input</code> step has an unique ID. It is used in the generated URL to proceed or abort.
  </p>
  <p>
    A specific ID could be used, for example, to mechanically respond the input from some external process/tool.
  </p>
<div>
```

Comment (recena) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: This wrapper (`div`) is very helpful if you want to do changes in all help boxes.

**Comment by jglick** [comment]
You might want to limit the scope of the PR so we can get independently useful stuff merged earlier.

File: basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStep.java [hunk 3]
```diff
@@ -61,7 +61,7 @@ public boolean takesImplicitBlockArgument() {
 
         @Override
         public String getDisplayName() {
-            return "Executes the body with a timeout";
+            return "Execution with timeout";
```
Comment (jglick) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStep.java]: Maybe `Enforce time limit`? (_Execution_ in English is pretty vague, even covering capital punishment!)

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/ArtifactArchiverStep/help-excludes.html [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<div>
+Exclude artifacts matching this <a href="https://ant.apache.org/manual/dirtasks.html#patterns">Ant style pattern</a>.
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/ArtifactArchiverStep/help-excludes.html]: BTW I recommend `target="_blank"` (IIRC) for all inline help, since Jenkins configuration pages tend to crash and burn if you return to them with a Back button.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/ArtifactArchiverStep/help-excludes.html [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+Use a comma separated list to add more than one expression.
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/ArtifactArchiverStep/help-excludes.html]: Also hyphenate `Ant-style` and `comma-separated`. (Helps avoid “garden path” parsing errors by readers.)

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<div>
+    <p>
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html]: Possibly OK but interactively check appearance of `<p>` to make sure there is not ugly extra padding.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+    This is a special step that allows to call build wrappers (also called "Environment Configuration" in freestyle or similar projects).
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html]: No, that is `CoreWrapperStep`. This one is for “build steps” and “post-build actions” (IIRC).

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep/help.html [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<div>
+    <p>
+    This is a special step that allows to call builders or post-build actions (as in freestyle or similar projects), in general "build steps".
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep/help.html]: Got these backwards.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PushdStep/help.html [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<div>
+    Change current directory. Any step inside the <code>dir</code> block will 
+    use this directory as current and any realtive path will use it as base path.
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PushdStep/help.html]: typo

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PwdStep/help.html [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<div>
+    Returns the current directory as a string.
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PwdStep/help.html]: Directory path, I would clarify.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PwdStep/help.html [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+<p><pre>
+def x = pwd()
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PwdStep/help.html]: Does this add anything beyond what snippet generation would say?

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/RetryStep/help.html [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Retry the body block up to N times if its execution finishes with status ABORTED or FAILURE.
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/RetryStep/help.html]: A block cannot have a “status” as such. (Pending [JENKINS-26522](https://issues.jenkins-ci.org/browse/JENKINS-26522) anyway.) Like a function in a programming language, which in fact it is, it can either return a value or throw an exception. Uncaught exceptions do get mapped onto various non-successful build statuses, but you can also set build status directly, and so on, so I think this is more misleading than helpful.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/TimeoutStep/help.html [hunk 13]
```diff
@@ -0,0 +1,4 @@
+<div>
+    Executes the code inside the block with a determined time out limit.
+    If the time limit is reached, the build is aborted.
```
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/TimeoutStep/help.html]: Technically, an exception is thrown which if uncaught will result in the build being aborted. Not sure if that is important enough to explain here, though.

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/Snippetizer/block.jelly [hunk 14]
```diff
@@ -28,7 +28,7 @@ THE SOFTWARE.
     <d:taglib uri="local">
         <d:tag name="listSteps">
             <j:forEach var="stepDescriptor" items="${stepDescriptors}">
-                <f:dropdownListBlock title="${stepDescriptor.displayName}" staplerClass="${stepDescriptor.clazz.name}" lazy="stepDescriptor,it">
+                <f:dropdownListBlock title="${stepDescriptor.functionName}: ${stepDescriptor.displayName}" staplerClass="${stepDescriptor.clazz.name}" lazy="stepDescriptor,it">
```
Comment (jglick) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/Snippetizer/block.jelly]: `Snippetizer.getStepDescriptors` will need to be modified to sort by `functionName`. (`ExtensionComponent.compareTo` sorts them in most cases by `displayName`, and `StepDescriptor.all` and thus currently `Snippetizer.getStepDescriptors` simply inherits that order.)

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help-path.html (hunk 2 above)
Comment (jglick) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help-path.html]: Ditto what?

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help.html [hunk 15]
```diff
@@ -0,0 +1,17 @@
+<div>
+    Takes a filename in the workspace and runs it as Groovy source text.
+    <p>
+        The loaded file can contain statements at top level or just load and run a closure. For example:
```
Comment (jglick) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help.html]: This is getting into the grey area of longer content, overlapping that described more thoroughly [in the “tutorial”](https://github.com/jenkinsci/workflow-plugin/blob/master/TUTORIAL.md#manual-loading). Cf. #203. I tend to recommend the full content be included here in `help.html`, where it will be available in other media such as via [JENKINS-26126](https://issues.jenkins-ci.org/browse/JENKINS-26126).

File: durable-task-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/durable_task/ShellStep/help.html [hunk 16]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Runs a Bourne shell script on a Unix node using <code>nohup</code>. Multiple lines are accepted.
```
Comment (jglick) [inline durable-task-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/durable_task/ShellStep/help.html]: The `nohup` aspect is probably confusing and irrelevant for most users, and anyway might become inaccurate if I ever get to [JENKINS-25503](https://issues.jenkins-ci.org/browse/JENKINS-25503).

Comment (jglick) [inline durable-task-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/durable_task/ShellStep/help.html]: What you *should* talk about is the fact that you may use

```perl
#!/usr/bin/perl
```

syntax to select an alternate interpreter, and also that you may want to run

```bash
set +e
```

and/or

```bash
set +x
```

to disable the flags which are set by default.

File: scm-step/src/main/java/org/jenkinsci/plugins/workflow/steps/scm/GitStep.java [hunk 17]
```diff
@@ -175,7 +175,7 @@ public FormValidation doCheckUrl(@AncestorInPath Item project,
         }
 
         @Override public String getDisplayName() {
-            return "Git";
+            return "SCM - Git";
```
Comment (jglick) [inline scm-step/src/main/java/org/jenkinsci/plugins/workflow/steps/scm/GitStep.java]: May not be necessary if sorting is fixed (see other comment).

File: scm-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/scm/GenericSCMStep/help.html [hunk 18]
```diff
@@ -0,0 +1,10 @@
+<div>
+    <p>
+    This is a special step that allows to call generic SCM steps. if you want to use a concrete SCM implementations,
```
Comment (jglick) [inline scm-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/scm/GenericSCMStep/help.html]: “Generic” probably not meaningful to the user. Should say this lets you run checkouts using any configuration options offered by any Workflow-compatible SCM plugin, and distinguish from the simplistic options offered by `git` and `svn` (which I should maybe just deprecate when we do [JENKINS-29922](https://issues.jenkins-ci.org/browse/JENKINS-29922) and fix default values for bean properties on `GitSCM` and `SubversionSCM`).

File: scm-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/scm/GitStep/help.html [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<div>
+    <p>
+    Git step. It performs a clone from the specified repository.
+    </p>
+    <p>
+    Note that this step is shorthand for the generic SCM step:<pre>
+checkout([$class: 'GitSCM', branches: [[name: '*/master']], 
+     userRemoteConfigs: [[url: 'http://git-server/user/repository.git']], ...])
```
Comment (jglick) [inline scm-step/src/main/resources/org/jenkinsci/plugins/workflow/steps/scm/GitStep/help.html]: No `...` needed here; it shorthand for the generic SCM step with no other options specified.

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help-concurrency.html [hunk 20]
```diff
@@ -0,0 +1,7 @@
+<p>
+    Concurrency level.
+</p>
+<p>
+    A concurrency of one is useful to let you lock a singleton resource, such as deployment to a single target server. 
+    Only one build will deploy at a given time: the newest which passed all previous stages
```
Comment (jglick) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help-concurrency.html]: Again I wonder if this should either link to the [tutorial](https://github.com/jenkinsci/workflow-plugin/blob/master/TUTORIAL.md#stages), or if the details there should be moved here.

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help.html [hunk 21]
```diff
@@ -0,0 +1,6 @@
+<div>
+    By default, flow builds can run concurrently. 
+    The stage command lets you mark certain sections of a build as being constrained by limited concurrency.
```
Comment (jglick) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help.html]: Should this text be moved into `help-concurrency.html`?

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html [hunk 22]
```diff
@@ -0,0 +1,6 @@
...
+    Every <code>input</code> step has an unique ID. It is used in the generated URL to proceed or abort.
+</p>
+<p>
+    A specific ID could be used, for example, to mechanically respond the input from some external process/tool.
```
Comment (jglick) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help-id.html]: “respond to the input”

**Comment by amuniz** [comment]
Yeah. let's split it (`work-in-progress` tag can be removed). Fixing comments...

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help.html [hunk 23]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Step that pauses for human input.
```
Comment (jglick) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/input/InputStep/help.html]: Please incorporate content from #203.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/ArtifactArchiverStep/help-excludes.html (hunk 5 above)
Comment (amuniz) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/ArtifactArchiverStep/help-excludes.html]: Ha! "garden path", interesting. English language corners.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html (hunk 6 above)
Comment (amuniz) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html]: I don't see anything weird in the UI.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PwdStep/help.html (hunk 11 above)
Comment (amuniz) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/PwdStep/help.html]: Not really. I wanted to show that the step itself does not perform anything internally, it only returns a value, but this is already clear from the sentence before, so... removing.

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help.html (hunk 15 above)
Comment (amuniz) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help.html]: Agree. But we can do that documentation move in a separate PR, ok?

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help.html (hunk 21 above)
Comment (amuniz) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help.html]: Well, I think this is general information, although `concurrency` has to be used inevitably here since it's the base of this step.

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/Snippetizer/block.jelly (hunk 14 above)
Comment (amuniz) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/Snippetizer/block.jelly]: Done.

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html (hunk 6 above)
Comment (amuniz) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html]: If you are referring to a padding inside the help box, yes, it is there, but it is outside the control of this plugin, so what's the correct way to go? A `<p>` tag here is the correct HTML markup, should I remove it because there is a CSS behaviour that should fixed somewhere else?

**Comment by amuniz** [comment]
Everything fixed (build failing unrelated to this pull request AFAICT)

**Comment by tfennelly** [comment]
:bee: great work @amuniz !!!

File: basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/ToolStep.java [hunk 24]
```diff
@@ -80,7 +80,7 @@ public DescriptorImpl() {
         }
 
         @Override public String getDisplayName() {
-            return "Install a tool";
+            return "Use a tool";
```
Comment (oleg-nenashev) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/ToolStep.java]: +/- "install" is a common term in Jenkins, BTW this term is also confusing

Comment (amuniz) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/ToolStep.java]: The `tool` step is not really installing the tool (it really depends on the existence of a `ToolInstaller`). That's why I think the term _Install_ here could be really confusing (actually, it was confusing to me before going to the code and see what the step execution really does).

Comment (oleg-nenashev) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/ToolStep.java]: I write something like `Use a tool from a predefined Tool Installation`. It should be more explicit for both users and Jenkins maintainers

Comment (amuniz) [inline basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/ToolStep.java]: Agree. Fixed.

**Comment by oleg-nenashev** [comment]
:bee:

File: basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html (hunk 6 above)
Comment (jglick) [inline basic-steps/src/main/resources/org/jenkinsci/plugins/workflow/steps/CoreStep/help.html]: No strong opinion.

File: cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help.html (hunk 15 above)
Comment (jglick) [inline cps/src/main/resources/org/jenkinsci/plugins/workflow/cps/steps/LoadStep/help.html]: Fine.

File: support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help.html (hunk 21 above)
Comment (jglick) [inline support/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/StageStep/help.html]: Most stages are used just for display purposes and do not specify `concurrency`.

File: cps/src/main/java/org/jenkinsci/plugins/workflow/cps/Snippetizer.java [hunk 25]
```diff
@@ -226,4 +229,12 @@ public HttpResponse doGenerateSnippet(StaplerRequest req, @QueryParameter String
         }
     }
 
+    private static class StepDescriptorComparator implements Comparator<StepDescriptor>, Serializable {
```
Comment (jglick) [inline cps/src/main/java/org/jenkinsci/plugins/workflow/cps/Snippetizer.java]: BTW you do not actually need to make this `Serializable` since it is only being used for rendering.

**Comment by jglick** [comment]
:bee:
